wallet-new/src/Cardano/Wallet/Kernel/Addresses.hs

```diff
@@ -0,0 +1,158 @@
+import           Cardano.Wallet.Kernel.DB.AcidState (CreateHdAddress (..))
+import           Cardano.Wallet.Kernel.DB.HdWallet (HdAccountId,
+                     HdAccountIx (..), HdAddressId (..), HdAddressIx (..),
+                     UnknownHdAccount (..), hdAccountIdIx, hdAccountIdParent,
+                     hdAddressIdIx)
+import           Cardano.Wallet.Kernel.DB.HdWallet.Create
+                     (CreateHdAddressError (..), initHdAddress)
+import           Cardano.Wallet.Kernel.DB.HdWallet.Derivation
+                     (HardeningMode (..), deriveIndex)
+import           Cardano.Wallet.Kernel.DB.InDb (InDb (..))
+import qualified Cardano.Wallet.Kernel.Keystore as Keystore
+import           Cardano.Wallet.Kernel.Types (AccountId (..), WalletId (..))
+
+import           Test.QuickCheck (Arbitrary (..), oneof)
+
+data CreateAddressError =
+      CreateAddressUnknownHdAccount UnknownHdAccount
+      -- ^ When trying to create the 'Address', the parent 'Account' was not
+      -- there.
+    | CreateAddressKeystoreNotFound AccountId
+      -- ^ When trying to create the 'Address', the 'Keystore' didn't have
+      -- any secret associated with this 'Account'.
+      -- there.
+    | CreateAddressHdCreationFailed CreateHdAddressError
+      -- ^ The creation of the new HD 'Address' failed with a database error.
+    | CreateAddressHdRndGenerationFailed HdAccountId
+      -- ^ The crypto-related part of address generation failed.
+    | CreateAddressHdRndAddressSpaceSaturated HdAccountId
+      -- ^ The available number of HD addresses in use in such that trying
+      -- to find another random index would be too expensive
+    deriving Eq
+
+-- TODO(adn)
+instance Arbitrary CreateAddressError where
+    arbitrary = oneof []
+
+instance Buildable CreateAddressError where
+    build (CreateAddressUnknownHdAccount uAccount) =
+        bprint ("CreateAddressUnknownHdAccount " % F.build) uAccount
+    build (CreateAddressKeystoreNotFound accId) =
+        bprint ("CreateAddressKeystoreNotFound " % F.build) accId
+    build (CreateAddressHdCreationFailed hdErr) =
+        bprint ("CreateAddressHdCreationFailed " % F.build) hdErr
+    build (CreateAddressHdRndGenerationFailed hdAcc) =
+        bprint ("CreateAddressHdRndGenerationFailed " % F.build) hdAcc
+    build (CreateAddressHdRndAddressSpaceSaturated hdAcc) =
+        bprint ("CreateAddressHdRndAddressSpaceSaturated " % F.build) hdAcc
+
+-- | Creates a new 'Address' for the input account.
+createAddress :: PassPhrase
+              -- ^ The 'Passphrase' (a.k.a the \"Spending Password\").
+              -> AccountId
+              -- ^ An abstract notion of an 'Account' identifier
+              -> PassiveWallet
+              -> IO (Either CreateAddressError Address)
+createAddress spendingPassword accId pw = do
+    let keystore = pw ^. walletKeystore
+    case accId of
+         -- \"Standard\" HD random derivation. The strategy is as follows:
+         --
```

Comment:
Reason 0: We cannot generate random indexes inside an acid-state transaction, this pretty much is sufficient condition to justify the rest of the code.